### PR TITLE
Data Factory Child Resource Update

### PR DESCRIPTION
**Pull request by senavar**
# Change

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update (Wiki)

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

[DataFactory: Factories](https://github.com/Azure/ResourceModules/actions/workflows/ms.datafactory.factories.yml)

Pipeline deploys successfully but the removal is failing:

### CODE REVIEW

File: arm/Microsoft.DataFactory/factories/.bicep/nested_rbac.bicep [hunk 1]
```diff
@@ -20,6 +20,7 @@ var builtInRoleNames = {
 }
 
 resource dataFactory 'Microsoft.DataFactory/factories@2018-06-01' = {
```
Comment (MrMCake) [inline arm/Microsoft.DataFactory/factories/.bicep/nested_rbac.bicep:22]: Rookie mistake ;) 

In bicep we use the concept of an `existing` resource to allow us to use it as a reference in files without actually deploying it.

What your solution should look like is this:
```bicep
resource dataFactory 'Microsoft.DataFactory/factories@2018-06-01' existing = {
  name: resourceName
}
```

If you'd convert this bad boy to JSON you'll notice that ARM converts this to a simple resource reference - i.e. it is not a resource deployment

File: arm/Microsoft.DataFactory/factories/deploy.bicep [hunk 2]
```diff
@@ -125,26 +134,21 @@ resource dataFactory 'Microsoft.DataFactory/factories@2018-06-01' = {
   }
 }
 
-resource dataFactory_managedVirtualNetwork 'Microsoft.DataFactory/factories/managedVirtualNetworks@2018-06-01' = {
-  parent: dataFactory
-  name: 'default'
-  properties: {}
+module dataFactory_managedVirtualNetwork 'managedVirtualNetwork/deploy.bicep' = {
```
Comment (MrMCake) [inline arm/Microsoft.DataFactory/factories/deploy.bicep:137]: Is this child mandatory? If not I'd suggest to add a condition to the resource such as:

```bicep
module dataFactory_managedVirtualNetwork 'managedVirtualNetwork/deploy.bicep' = if (!empty(managedVirtualNetworkName)) {
```

Comment (MrMCake) [reply arm/Microsoft.DataFactory/factories/deploy.bicep:137]: On of course make the parameter non-mandatory in this file too. However, do this only if the resource is actually not required

File: arm/Microsoft.DataFactory/factories/deploy.bicep [hunk 3]
```diff
@@ -125,26 +134,21 @@ resource dataFactory 'Microsoft.DataFactory/factories@2018-06-01' = {
...
-    managedVirtualNetwork: {
-      referenceName: 'default'
-      type: 'ManagedVirtualNetworkReference'
-    }
-    typeProperties: {
-      computeProperties: {
-        location: 'AutoResolve'
-      }
-    }
+module dataFactory_integrationRuntime 'integrationRuntime/deploy.bicep' = {
```
Comment (MrMCake) [inline arm/Microsoft.DataFactory/factories/deploy.bicep:145]: Same question as before: Is this resource mandatory?

File: arm/Microsoft.DataFactory/factories/integrationRuntime/readme.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
+# DataFactoryIntegrationRunTimes `[Microsoft.DataFactory/factories/integrationRuntimes]`
```
Comment (MrMCake) [inline arm/Microsoft.DataFactory/factories/integrationRuntime/readme.md:1]: Please update to a meaningful name like `Data Factory Integration Runtime` and add a description below. Can be something easy like `This module deploys a data factory integration runtime`

File: arm/Microsoft.DataFactory/factories/integrationRuntime/readme.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+| Parameter Name | Type | Default Value | Possible Values | Description |
+| :-- | :-- | :-- | :-- | :-- |
+| `cuaId` | string |  |  | Optional. Customer Usage Attribution ID (GUID). This GUID must be previously registered|
+| `name` | string |  |  | Required. The name of the Integration Runtime. |
+| `dataFactoryName` | string |  |  | Required. The name of the Azure Factory to create |
+| `managedVirtualNetworkName` | string |  |  | Required. The name of the Managed Virtual Network. |
+| `typeProperties` | object |  |  | Required. Managed integration runtime type properties. |
+
+### Parameter Usage: `typeProperties`
+
```
Comment (MrMCake) [inline arm/Microsoft.DataFactory/factories/integrationRuntime/readme.md:20]: This is a very complicated object. Please add this [reference](https://docs.microsoft.com/en-us/azure/templates/microsoft.datafactory/factories/integrationruntimes?tabs=bicep#integrationruntime-objects) here:

File: arm/Microsoft.DataFactory/factories/managedVirtualNetwork/readme.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
+# DataFactoryManagedVirtualNetwork `[Microsoft.DataFactory/factories/managedVirtualNetworks]`
```
Comment (MrMCake) [inline arm/Microsoft.DataFactory/factories/managedVirtualNetwork/readme.md:1]: Same as before, please update to a more meaningful name & add a description

File: arm/Microsoft.DataFactory/factories/deploy.bicep [hunk 7]
```diff
@@ -125,26 +134,21 @@ resource dataFactory 'Microsoft.DataFactory/factories@2018-06-01' = {
...
+  name: '${uniqueString(deployment().name, location)}-IntegrationRuntime'
+  params: {
+    name: integrationRuntimeName
+    dataFactoryName: dataFactory.name
+    managedVirtualNetworkName: managedVirtualNetworkName
+    typeProperties: typeProperties
```
Comment (MrMCake) [inline arm/Microsoft.DataFactory/factories/deploy.bicep:151]: Please move all integrationRuntime properties into a parameter `param integrationRuntime object`. We should try to keep child resource properties away from the parent.
Moving it to an object allows us to update the child without updating the parent parameters again.
After updating please re-run the readme generation script for the parent readme

File: arm/Microsoft.DataFactory/factories/.bicep/nested_rbac.bicep (hunk 1 above)
Comment (senavar) [reply arm/Microsoft.DataFactory/factories/.bicep/nested_rbac.bicep]: Thank you for calling this out! Done for the Data Factory. We may have to go through modules and see if this is occurring on others :)

File: arm/Microsoft.DataFactory/factories/deploy.bicep (hunk 2 above)
Comment (senavar) [reply arm/Microsoft.DataFactory/factories/deploy.bicep]: Not mandatory, this has been fixed

File: arm/Microsoft.DataFactory/factories/deploy.bicep (hunk 3 above)
Comment (senavar) [reply arm/Microsoft.DataFactory/factories/deploy.bicep]: Not mandatory, this has been fixed

File: arm/Microsoft.DataFactory/factories/integrationRuntime/readme.md (hunk 4 above)
Comment (senavar) [reply arm/Microsoft.DataFactory/factories/integrationRuntime/readme.md]: Updated!

File: arm/Microsoft.DataFactory/factories/integrationRuntime/readme.md (hunk 5 above)
Comment (senavar) [reply arm/Microsoft.DataFactory/factories/integrationRuntime/readme.md:22]: Added!

File: arm/Microsoft.DataFactory/factories/managedVirtualNetwork/readme.md (hunk 6 above)
Comment (senavar) [reply arm/Microsoft.DataFactory/factories/managedVirtualNetwork/readme.md]: Updated!

File: arm/Microsoft.DataFactory/factories/deploy.bicep (hunk 7 above)
Comment (senavar) [reply arm/Microsoft.DataFactory/factories/deploy.bicep]: This has been updated. There is now an integrationRuntime object and have added the capability of deploying a managed or self-hosted Integration Runtime

File: arm/Microsoft.DataFactory/factories/deploy.bicep [hunk 8]
```diff
@@ -125,26 +131,22 @@ resource dataFactory 'Microsoft.DataFactory/factories@2018-06-01' = {
...
-        location: 'AutoResolve'
-      }
-    }
+module dataFactory_integrationRuntime 'integrationRuntime/deploy.bicep' = if (!empty(integrationRuntime)) {
+  name: '${uniqueString(deployment().name, location)}-IntegrationRuntime'
+  params: {
+    dataFactoryName: dataFactory.name
+    name: integrationRuntime.name
+    type: integrationRuntime.type
+    managedVirtualNetworkName: integrationRuntime.managedVirtualNetworkName
```
Comment (MrMCake) [inline arm/Microsoft.DataFactory/factories/deploy.bicep:148]: As this value seems optional you should handle it as such: `managedVirtualNetworkName: contains(integrationRuntime, 'managedVirtualNetworkName') ? integrationRuntime.managedVirtualNetworkName : ''`

File: arm/Microsoft.DataFactory/factories/readme.md [hunk 9]
```diff
@@ -35,6 +37,21 @@
 | `tags` | object | `{object}` |  | Optional. Tags of the resource. |
 | `workspaceId` | string |  |  | Optional. Resource identifier of Log Analytics. |
 
+
+### Parameter Usage: [`integrationRuntime`](https://docs.microsoft.com/en-us/azure/templates/microsoft.datafactory/factories/integrationruntimes?tabs=bicep)
```
Comment (MrMCake) [inline arm/Microsoft.DataFactory/factories/readme.md:41]: Can be removed. If you generate the ReadMe with the function, the parameter should get a link in the above parameters table type column. Please doublecheck

File: arm/Microsoft.DataFactory/factories/deploy.bicep (hunk 8 above)
Comment (senavar) [reply arm/Microsoft.DataFactory/factories/deploy.bicep]: Updated!

File: arm/Microsoft.DataFactory/factories/readme.md (hunk 9 above)
Comment (senavar) [reply arm/Microsoft.DataFactory/factories/readme.md]: Unfortunately, for some reason the ReadMe script does now work for me. But that's a conversation for another thread :)
image

I have gone and updated manually based on what's in other modules

Comment (MrMCake) [reply arm/Microsoft.DataFactory/factories/readme.md]: Added a minor update in PR : https://github.com/Azure/ResourceModules/pull/529

File: arm/Microsoft.DataFactory/factories/deploy.bicep (hunk 8 above)
Comment (MrMCake) [reply arm/Microsoft.DataFactory/factories/deploy.bicep]: Just one minor change: Please remove the ecapuslating brackets ( `(` ) they are redundant

Comment (senavar) [reply arm/Microsoft.DataFactory/factories/deploy.bicep]: Done!

**Review by MrMCake — APPROVED** [review]

**Review by MrMCake — APPROVED** [review]

**Pull request merged**